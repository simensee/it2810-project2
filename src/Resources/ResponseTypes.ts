
// Autogenerated from https://app.quicktype.io/

export interface User {
    id?:        number;
    username?:  string;
    name:       string;
    state?:     string;
    avatar_url?: string;
    web_url?:    string;
    founded?:   number;
    members?:   string[];
}

export interface Action {
    id?:             number;
    projectID?:      number;
    actionName?:     string;
    targetID?:       null;
    targetIid?:      null;
    targetType?:     null;
    authorID?:       number;
    targetTitle?:    null;
    createdAt?:      Date;
    author?:         Author;
    pushData?:       PushData;
    authorUsername?: string;
    name?:           string;
    founded?:        number;
    members?:        string[];
}

export interface Author {
    id:        number;
    username:  string;
    name:      string;
    state:     string;
    avatarURL: string;
    webURL:    string;
}

export interface PushData {
    commitCount: number;
    action:      string;
    refType:     string;
    commitFrom:  string;
    commitTo:    string;
    ref:         string;
    commitTitle: string;
    refCount:    null;
}

export interface Commit {
    id:              string;
    short_id:        string;
    created_at:      Date;
    parent_ids:      string[];
    title:           string;
    message:         string;
    author_name:     string;
    author_email:    string;
    authored_date:   Date;
    committer_name:  string;
    committer_email: string;
    committed_date:  Date;
    trailers:       boolean;
    web_url:         string;
}

export interface MergeRequest {
    id?:                          number;
    iid?:                         number;
    projectID?:                   number;
    title?:                       string;
    description?:                 string;
    state?:                       string;
    createdAt?:                   Date;
    updatedAt?:                   Date;
    mergedBy?:                    User;
    mergeUser?:                   User;
    mergedAt?:                    Date;
    closedBy?:                    null;
    closedAt?:                    null;
    targetBranch?:                string;
    sourceBranch?:                string;
    userNotesCount?:              number;
    upvotes?:                     number;
    downvotes?:                   number;
    author?:                      User;
    assignees?:                   User[];
    assignee?:                    User;
    reviewers?:                   User[];
    sourceProjectID?:             number;
    targetProjectID?:             number;
    labels?:                      string[];
    draft?:                       boolean;
    workInProgress?:              boolean;
    milestone?:                   null;
    mergeWhenPipelineSucceeds?:   boolean;
    mergeStatus?:                 string;
    sha?:                         string;
    mergeCommitSHA?:              string;
    squashCommitSHA?:             null;
    discussionLocked?:            null;
    shouldRemoveSourceBranch?:    boolean;
    forceRemoveSourceBranch?:     boolean;
    reference?:                   string;
    references?:                  References;
    webURL?:                      string;
    timeStats?:                   TimeStats;
    squash?:                      boolean;
    taskCompletionStatus?:        TaskCompletionStatus;
    hasConflicts?:                boolean;
    blockingDiscussionsResolved?: boolean;
}

export interface References {
    short?:    string;
    relative?: string;
    full?:     string;
}

export interface TaskCompletionStatus {
    count?:          number;
    completedCount?: number;
}

export interface TimeStats {
    timeEstimate?:        number;
    totalTimeSpent?:      number;
    humanTimeEstimate?:   null;
    humanTotalTimeSpent?: null;
}
