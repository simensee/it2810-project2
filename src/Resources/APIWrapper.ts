import { useQuery } from "@tanstack/react-query";
import React from "react";
import { Action, User } from "./ResponseTypes";

const stringUrl: string = 'https://gitlab.stud.idi.ntnu.no/api/v4/projects/17475/';
const APIToken: string = 'glpat-cygbLETJKv1wXaNyMtXS';


export function FetchUsers(): User[] {
    const usersUrl = stringUrl.concat('users');
    let fetchUsersUrl: URL = new URL(usersUrl);
    const {isLoading, isError, data, error} = useQuery(['userData'], () => 
        fetch(fetchUsersUrl, {
            method: 'GET', 
            headers: new Headers({
                'Content-Type': 'application/json',
                'Authorization': 'Bearer '+ APIToken,
            })}).then(res => res.json())
    );

    console.log(data);

    if (isLoading) console.log('fetching');    

    // Does not catch 40x responses
    if (isError) {
        console.log('An error occurred '+ error);
    } 
    const response: User[] = data;
    return response;
}

export function FetchEvents(): Action[] {
    const usersUrl = stringUrl.concat('events');
    let fetchUsersUrl: URL = new URL(usersUrl);
    const {isLoading, isError, data, error} = useQuery(['eventData'], () => 
        fetch(fetchUsersUrl, {
            method: 'GET', 
            headers: new Headers({
                'Content-Type': 'application/json',
                'Authorization': 'Bearer '+ APIToken,
            })}).then(res => res.json())
    );

    if (isLoading) console.log('fetching');    

    // Does not catch 40x responses
    if (isError) {
        console.log('An error occurred '+ error);
    } 
    
    const response: Action[] = data;   
    return response;
}
