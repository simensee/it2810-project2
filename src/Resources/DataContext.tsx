import React, { createContext, useState, ReactNode } from 'react'
import { Commit, User } from './ResponseTypes';


interface DataContextProps {
    usersData: User[],
    commitData: Commit[],
    fetchUsers: Function,
    fetchCommits: Function,
}

export const DataContext = createContext<DataContextProps>({
    usersData: [],
    commitData: [],
    fetchUsers: () => null,
    fetchCommits: () => null,
});

export interface LayoutProps {
    children: ReactNode,
}

export const DataContextProvider = (props: LayoutProps) => {
    const stringUrl: string = 'https://gitlab.stud.idi.ntnu.no/api/v4/projects/17475/';
    const APIToken: string = 'glpat-cygbLETJKv1wXaNyMtXS';

    let usersData: User[] = [];
    let commitData: Commit[] = [];

    const fetchUsers = async () => {
        const usersUrl = stringUrl.concat('users');
        let fetchUsersUrl: URL = new URL(usersUrl);

        await fetch(fetchUsersUrl, {
            method: 'GET',
            headers: new Headers({
                'Content-Type': 'application/json',
                'Authorization': 'Bearer ' + APIToken,
            })
        }).then(res => res.json()).then((res) => {
            const userResponse: User[] = res;
            usersData.push(...userResponse);
        });
    }

    const fetchCommits = async () => {
        const commitUrl = stringUrl.concat('repository/commits');
        let fetchCommitUrl: URL = new URL(commitUrl);

        await fetch(fetchCommitUrl, {
            method: 'GET',
            headers: new Headers({
                'Content-Type': 'application/json',
                'Authorization': 'Bearer ' + APIToken,
            })
        }).then(res => res.json()).then((res) => {
            const commitResponse: Commit[] = res;
            commitData.push(...commitResponse);
        });
    }

    

    const value: DataContextProps = {
        usersData,
        commitData,
        fetchUsers,
        fetchCommits,
    };

    return <DataContext.Provider value={value}>{props.children}</DataContext.Provider>
}
