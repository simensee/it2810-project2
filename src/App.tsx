import React, { useContext, useEffect, useState } from 'react';
import logo from './logo.svg';
import { Commit, User } from './Resources/ResponseTypes';
import Sidebar from './Components/Sidebar/Sidebar';
import NavRouter from './Components/Router/NavRouter';
import { DataContext } from './Resources/DataContext';
import { classicNameResolver } from 'typescript';
import classnames from 'tailwindcss-classnames';
import LoginPage from './Components/Pages/LoginPage';

function App() {
  const ctx = useContext(DataContext);

  const setup = async () => {
    await ctx.fetchUsers();
    await ctx.fetchCommits();
  }

  const [loadComplete, setLoadComplete] = useState(false);

  useEffect(() => {
    setup().then(() => setLoadComplete(true)).then(() => console.log(ctx.usersData));
  }, []);

  // console.log(FetchUsers());

  // console.log(resp[0]);s


  // resp?.map(u => console.log(u.name))
  // console.log(FetchEvents());

  // ctx.setVal('Hallo');

  return (
    <>
      {/* <LoginPage/> */}
      {loadComplete ? 
      <NavRouter/>
      :
      <div className='flex items-center justify-center h-full'>
        <span className='text-5xl'>Loading...</span>
      </div>      
    }
      {/* <nav>
        <Sidebar />
      </nav>
      <div className='ml-64 pl-4 pb-8 lg:pt-12 px-4 h-full'>
        {loadComplete ?
          <span>done</span>
          : <span>Loading data</span>
        }
        <NavRouter/>
      </div> */}
    </>
  );
}

export default App;
