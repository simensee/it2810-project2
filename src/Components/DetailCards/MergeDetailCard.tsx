import React, { useContext, useEffect, useState } from 'react'
import { DataContext } from '../../Resources/DataContext';
import { Commit, MergeRequest } from '../../Resources/ResponseTypes';


type MergeDetailCardProps = {
    focusMerge: MergeRequest;
}

const MergeDetailCard = ({ focusMerge }: MergeDetailCardProps) => {

    const ctx = useContext(DataContext);
    //const [userCommits, setUserCommits] = useState(0);

   

    return (
        <div>
            {focusMerge.id}
        </div>
    )
}

export default MergeDetailCard