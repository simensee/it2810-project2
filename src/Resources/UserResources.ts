import { useContext } from "react";
import { DataContext } from "./DataContext";
import { Commit, User } from "./ResponseTypes";


// const ctx = useContext(DataContext);

