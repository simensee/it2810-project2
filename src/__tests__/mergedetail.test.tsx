import React, { useContext } from "react";
import App from "../App";
import renderer from "react-test-renderer";
import {DataContext} from "../Resources/DataContext";
import { MergeRequest, User } from "../Resources/ResponseTypes";
import MergeDetailCard from "../Components/DetailCards/MergeDetailCard";



 

 describe('Merge component', () => {
   it('it works', async () => {
     let mergeRequest : MergeRequest = ({
         id: 62388
     });

      
    const tree : any = renderer.create(<MergeDetailCard focusMerge={mergeRequest} />);
    
     
    expect(tree.toJSON()).toMatchSnapshot()
    
 });
});

