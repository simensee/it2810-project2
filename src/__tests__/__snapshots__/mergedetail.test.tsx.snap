// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component it works 1`] = `
<div>
  62388
</div>
`;
